generator/api/src/test/java/io/fabric8/maven/generator/api/FromSelectorTest.java

```diff
@@ -97,26 +97,92 @@ public void simpleJava11() {
+                openshift, docker, "1.2.3.foo-00009", "docker-prop-8", "istag-prop",
+                openshift, s2i, "1.2.3", "s2i-prop", "istag-prop",
+                openshift, docker, "1.2.3", "docker-prop-8", "istag-prop",
+                null, s2i, "1.2.3.redhat-00009", "redhat-docker-prop", "redhat-istag-prop",
+                null, docker, "1.2.3.redhat-00009", "redhat-docker-prop", "redhat-istag-prop",
+                null, s2i, "1.2.3.fuse-00009", "redhat-docker-prop", "redhat-istag-prop",
+                null, docker, "1.2.3.fuse-00009", "redhat-docker-prop", "redhat-istag-prop",
+                null, s2i, "1.2.3.foo-00009", "docker-prop-8", "istag-prop",
+                null, docker, "1.2.3.foo-00009", "docker-prop-8", "istag-prop",
+                null, s2i, "1.2.3", "docker-prop-8", "istag-prop",
+                null, docker, "1.2.3", "docker-prop-8", "istag-prop",
+                openshift, null, "1.2.3.redhat-00009", "redhat-docker-prop", "redhat-istag-prop",
+                openshift, null, "1.2.3.fuse-00009", "redhat-docker-prop", "redhat-istag-prop",
+                openshift, null, "1.2.3.foo-00009", "docker-prop-8", "istag-prop",
+                openshift, null, "1.2.3", "docker-prop-8", "istag-prop"
+        };
+
+        for (int i = 0; i < data.length; i += 5) {
+            prepareExpectionJava8Target((String) data[i + 2], (RuntimeMode) data[i], (OpenShiftBuildStrategy) data[i + 1]);
+            FromSelector selector = new FromSelector.Default(ctx, "test");
+
+            assertEquals(data[i + 3], selector.getFrom());
+            Map<String, String> fromExt = selector.getImageStreamTagFromExt();
+            assertEquals(fromExt.size(),3);
+            assertEquals(fromExt.get(SourceStrategy.kind.key()), "ImageStreamTag");
+            assertEquals(fromExt.get(SourceStrategy.namespace.key()), "openshift");
+            assertEquals(fromExt.get(SourceStrategy.name.key()), data[i + 4]);
+        }
+    }
+
+    @Test
+    public void Java11BaseImageTarget() {
+        final Object[] data = new Object[] {
+                openshift, docker, "1.2.3.foo-00009", "docker-prop-11",
+                openshift, docker, "1.2.3", "docker-prop-11",
+                null, s2i, "1.2.3.foo-00009", "docker-prop-11",
+                null, docker, "1.2.3.foo-00009", "docker-prop-11",
+                openshift, null, "1.2.3.foo-00009", "docker-prop-11",
+                openshift, null, "1.2.3", "docker-prop-11",
+        };
+
+        for (int i = 0; i < data.length; i += 4) {
+            prepareExpectionJava11Target((String) data[i + 2], (RuntimeMode) data[i], (OpenShiftBuildStrategy) data[i + 1]);
+            FromSelector selector = new FromSelector.Default(ctx, "test");
+
+            assertEquals(data[i + 3], selector.getFrom());
+        }
+    }
+
+    private Expectations prepareExpectionJava8Release(final String version, final RuntimeMode mode, final OpenShiftBuildStrategy strategy) {
```

Comment:
All of these four methods seem to be doing the same thing. Do we really need this just for the sake of improving readability? Wouldn't the test name give us an idea about java release type being tested?

Reply:
these 4 prepare functions call the main prepare function prepareExpectation with 4 different combinations of properties...java version and release/target property...which is to checked by the FromSelector class.

Reply:
Yes, I can see that but I don't see their significance.

Reply:
they prepare the expections for JMockit for different config scenarios

Reply:
I think simply calling prepareExpectations directly would also achieve this. 

Anyway it's not something super critical.

Reply:
exactly